# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

---

## Step 1

In [1]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv"
df = pd.read_csv(url)

# Check the first few rows of the dataset
df.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [2]:
# Filter the DataFrame
filtered_df = df[(df['Total Claim Amount'] > 1000) & (df['Response'] == 'Yes')]

# Display the filtered DataFrame
filtered_df.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
189         189  OK31456  California             11009.130490      Yes   
236         236  YJ16163      Oregon             11009.130490      Yes   
419         419  GW43195      Oregon             25807.063000      Yes   
442         442  IP94270     Arizona             13736.132500      Yes   
587         587  FJ28407  California              5619.689084      Yes   

     Coverage             Education Effective To Date EmploymentStatus Gender  \
189   Premium              Bachelor           1/24/11         Employed      F   
236   Premium              Bachelor           1/24/11         Employed      F   
419  Extended               College           2/13/11         Employed      F   
442   Premium                Master           2/13/11         Disabled      F   
587   Premium  High School or Below           1/26/11       Unemployed      M   

     ...  Number of Open Complaints Number of Policies     Policy Type  \
189  ...                        0.0                  1  Corporate Auto   
236  ...                        0.0                  1    Special Auto   
419  ...                        1.0                  2   Personal Auto   
442  ...                        0.0                  8   Personal Auto   
587  ...                        0.0                  1   Personal Auto   

           Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
189  Corporate L3            Offer2          Agent         1358.400000   
236    Special L3            Offer2          Agent         1358.400000   
419   Personal L2            Offer1         Branch         1027.200000   
442   Personal L2            Offer1            Web         1261.319869   
587   Personal L1            Offer2            Web         1027.000029   

     Vehicle Class Vehicle Size Vehicle Type  
189     Luxury Car      Medsize          NaN  
236     Luxury Car      Medsize            A  
419     Luxury Car        Small            A  
442            SUV      Medsize            A  
587            SUV      Medsize            A  

[5 rows x 26 columns]

---

## Step 2

In [8]:
# Filter the original DataFrame for customers who responded "Yes"
yes_response_df = df[df['Response'] == 'Yes']

# Group by policy_type and gender, then calculate the mean of total_claim_amount
average_claims = yes_response_df.groupby(['Policy Type', 'Gender'])['Total Claim Amount'].mean().reset_index()

# Display the result
average_claims

Policy Type Gender  Total Claim Amount
0  Corporate Auto      F          433.738499
1  Corporate Auto      M          408.582459
2   Personal Auto      F          452.965929
3   Personal Auto      M          457.010178
4    Special Auto      F          453.280164
5    Special Auto      M          429.527942

### Conclusions:

1. **Gender Differences**:
   - For **Corporate Auto** and **Special Auto** policies, **females** have a slightly higher average claim amount than **males**.
   - However, for **Personal Auto** policies, the trend reverses slightly, with **males** having a slightly higher average claim amount than **females**.

2. **Policy Type Differences**:
   - The **Personal Auto** and **Special Auto** policies have **higher average claim amounts** compared to **Corporate Auto**.
   - Both genders generally have similar average claims for each policy type (with small differences), but the exact difference may still be significant depending on your analytical focus.

---

## Step 3:

In [11]:
# Group by state and count the number of customers
state_counts = df['State'].value_counts().reset_index()

# Rename columns for clarity
state_counts.columns = ['State', 'Customer Count']

# Display the result
state_counts

State  Customer Count
0  California            3552
1      Oregon            2909
2     Arizona            1937
3      Nevada             993
4  Washington             888

In [ ]:
# Filter to include only states with more than 500 customers (unnecessary, but for the sake of the example, we'll include it here)
filtered_states = state_counts[state_counts['Customer Count'] > 500]

# Display the filtered result (wow, all States are over 500 customers, who would have thought that!)
filtered_states.head()

State  Customer Count
0  California            3552
1      Oregon            2909
2     Arizona            1937
3      Nevada             993
4  Washington             888

---

## Step 4

In [15]:
# Group by education level and gender
clv_stats = df.groupby(['Education', 'Gender'])['Customer Lifetime Value'].agg(['max', 'min', 'median']).reset_index()

# Display the result
clv_stats

Education Gender          max          min       median
0              Bachelor      F  73225.95652  1904.000852  5640.505303
1              Bachelor      M  67907.27050  1898.007675  5548.031892
2               College      F  61850.18803  1898.683686  5623.611187
3               College      M  61134.68307  1918.119700  6005.847375
4                Doctor      F  44856.11397  2395.570000  5332.462694
5                Doctor      M  32677.34284  2267.604038  5577.669457
6  High School or Below      F  55277.44589  2144.921535  6039.553187
7  High School or Below      M  83325.38119  1940.981221  6286.731006
8                Master      F  51016.06704  2417.777032  5729.855012
9                Master      M  50568.25912  2272.307310  5579.099207

### Conclusions

1. **Gender Differences**:
   - Across most education levels, **females** tend to have a slightly higher **maximum CLV**, except for **High School or Below**, where **males** reach the highest overall max ($83,325.38).
   - **Median CLVs** are fairly close between genders but are often higher for **males**, particularly in **College** and **High School or Below**.

2. **Education Differences**:
   - Surprisingly, the **highest maximum CLV** comes from customers with **High School or Below** education, not higher degrees.
   - **Median CLVs** tend to be higher in lower or mid-range education levels like **College** and **High School or Below**, suggesting customer lifetime value does not increase linearly with education.

3. **Insights**:
   - There's no direct correlation between higher education and higher customer lifetime value.
   - Customers with **High School or Below** education may represent a high-value segment worth analyzing further.

---

## Step 5

In [17]:
# Check if the column 'Effective To Date' is in datetime format
df['Effective To Date'] = pd.to_datetime(df['Effective To Date'])

# Extract the month name from the date
df['Month'] = df['Effective To Date'].dt.strftime('%B')

# Create the pivot table
policy_pivot = df.pivot_table(index='State', 
                               columns='Month', 
                               values='Policy', 
                               aggfunc='count',
                               fill_value=0)

# Optional: sort months in calendar order
month_order = ['January', 'February', 'March', 'April', 'May', 'June',
               'July', 'August', 'September', 'October', 'November', 'December']
policy_pivot = policy_pivot.reindex(columns=month_order)

# Display the pivot table
policy_pivot

Month       January  February  March  April  May  June  July  August  \
State                                                                  
Arizona        1008       929    NaN    NaN  NaN   NaN   NaN     NaN   
California     1918      1634    NaN    NaN  NaN   NaN   NaN     NaN   
Nevada          551       442    NaN    NaN  NaN   NaN   NaN     NaN   
Oregon         1565      1344    NaN    NaN  NaN   NaN   NaN     NaN   
Washington      463       425    NaN    NaN  NaN   NaN   NaN     NaN   

Month       September  October  November  December  
State                                               
Arizona           NaN      NaN       NaN       NaN  
California        NaN      NaN       NaN       NaN  
Nevada            NaN      NaN       NaN       NaN  
Oregon            NaN      NaN       NaN       NaN  
Washington        NaN      NaN       NaN       NaN

---

## Step 6

In [18]:
# 1: Group by state and count total number of policies sold
total_policies_by_state = df.groupby('State')['Policy'].count().sort_values(ascending=False)

# 2: Identify the top 3 states with the most policies sold
top_3_states = total_policies_by_state.head(3).index.tolist()

# 3: Filter the DataFrame to include only those top 3 states
filtered_df = df[df['State'].isin(top_3_states)]

# 4: Group by state and month, count policies sold
top_states_monthly_sales = filtered_df.groupby(['State', 'Month'])['Policy'].count().reset_index()

# 5: Pivot the table to show months as columns and states as rows
final_df = top_states_monthly_sales.pivot(index='State', columns='Month', values='Policy').fillna(0)

# Optional: Reorder columns to match calendar order
month_order = ['January', 'February', 'March', 'April', 'May', 'June',
               'July', 'August', 'September', 'October', 'November', 'December']
final_df = final_df.reindex(columns=month_order)

# Display final DataFrame
final_df

Month       January  February  March  April  May  June  July  August  \
State                                                                  
Arizona        1008       929    NaN    NaN  NaN   NaN   NaN     NaN   
California     1918      1634    NaN    NaN  NaN   NaN   NaN     NaN   
Oregon         1565      1344    NaN    NaN  NaN   NaN   NaN     NaN   

Month       September  October  November  December  
State                                               
Arizona           NaN      NaN       NaN       NaN  
California        NaN      NaN       NaN       NaN  
Oregon            NaN      NaN       NaN       NaN

---

## Step 7

In [27]:
# Step 1: Filter to only relevant columns
channel_response = df[['Sales Channel', 'Response']]

# Step 2: Group by sales channel and response, count each combination
response_counts = channel_response.groupby(['Sales Channel', 'Response']).size().unstack(fill_value=0)

# Step 3: Calculate the response rate (Yes / total)
response_counts['response_rate'] = (
    response_counts['Yes'] / (response_counts['Yes'] + response_counts['No'])
).round(3) * 100  # convert to percentage

# Display result
response_counts[['response_rate']]

Response       response_rate
Sales Channel               
Agent                   19.1
Branch                  11.4
Call Center             11.0
Web                     11.7